benchmark/single-source/StringReplaceSubrange.swift

```diff
@@ -0,0 +1,116 @@
+//===--- StringReplaceSubrange.swift -------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2019 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import TestsUtils
+
+let tags: [BenchmarkCategory] = [.validation, .api, .String]
+
+public let StringReplaceSubrange = [
+  BenchmarkInfo(
+    name: "Str.replaceSubrange.SmallLiteral.String",
+    runFunction: { replaceSubrange($0, "coffee", with: "t") },
+    tags: tags
+  ),
+  BenchmarkInfo(
+    name: "Str.replaceSubrange.LargeManaged.String",
+    runFunction: { replaceSubrange($0, largeManagedString, with: "t") },
+    tags: tags,
+    setUpFunction: setupLargeManagedString
+  ),
+  BenchmarkInfo(
+    name: "Str.replaceSubrange.SmallLiteral.Substr",
+    runFunction: { replaceSubrange($0, "coffee", with: getSubstring("t")) },
+    tags: tags
+  ),
+  BenchmarkInfo(
+    name: "Str.replaceSubrange.LargeManaged.Substr",
+    runFunction: { replaceSubrange($0, largeManagedString, with: getSubstring("t")) },
+    tags: tags,
+    setUpFunction: setupLargeManagedString
+  ),
+  BenchmarkInfo(
+    name: "Str.replaceSubrange.SmallLiteral.ArrChar",
+    runFunction: { replaceSubrange($0, "coffee", with: getArrayCharacter(Array<Character>(["t"]))) },
+    tags: tags
+  ),
+  BenchmarkInfo(
+    name: "Str.replaceSubrange.LargeManaged.ArrChar",
+    runFunction: { replaceSubrange($0, largeManagedString, with: getArrayCharacter(Array<Character>(["t"]))) },
+    tags: tags,
+    setUpFunction: setupLargeManagedString
+  ),
+  BenchmarkInfo(
+    name: "Str.replaceSubrange.SmallLiteral.RepeatedChar",
+    runFunction: { replaceSubrange($0, "coffee", with: getRepeatedCharacter(repeatedCharacter)) },
+    tags: tags
```

Comment:
The benchmark name "Str.replaceSubrange.SmallLiteral.RepeatedChar" is longer than 40 characters, but I couldn't think a better name fitting 40. Maybe it can be like "Str.replaceSubrange.LargeManagedRepChar", but I was concerned "RepChar" is a little bit hard to understand that it means `Repeated<Character>`. @palimondo What do you think about this naming? Please let me know if you have any suggestions on it 🙂

Reply:
Looking at what @milseman writes in [SR-8905](https://bugs.swift.org/browse/SR-8905):

> replaceSubrange<C: Collection>(_:C)
> * Arguments of types `String`, `Substring`, `Array<Character>`, `Repeated<Character>`, etc

I'd say the [naming convention](http://bit.ly/BenchmarkNaming) calls for base name of `String.replaceSubrange` which varies across the argument type (`String`, `Substring`, `ArrChar`, `RepChar`)  for the general case of large strings. Then we'll denote the special optimization for small strings with a simple `.Small` suffix and we'll get these benchmarks:

* `String.replaceSubrange.String`
* `String.replaceSubrange.Substring`
* `String.replaceSubrange.ArrChar`
* `String.replaceSubrange.RepChar`
* `String.replaceSubrange.String.Small`
* `String.replaceSubrange.Substring.Small`
* `String.replaceSubrange.ArrChar.Small`
* `String.replaceSubrange.RepChar.Small`

The longest one is `String.replaceSubrange.Substring.Small` at 39 characters, just under the 40 chars limit.

Reply:
That's an awesome naming idea. I will use them. Thanks for your suggestion!